src/tests/suite/daos_obj.c

```diff
@@ -3792,6 +3792,85 @@ io_invalid(void **state)
 	assert_int_equal(rc, 0);
 }
 
+/**
+ * Write multiple 4k I/O extents with start index 0, then again
+ * with start index 1, with an index gap between each record,
+ * then wait for aggregation to be triggered.
+ * The gap in indexes prevents records from combining and allows for
+ * maximum amount of read I/Os upon aggregation.
+ */
+static void
+max_io_agg_threshold_internal(daos_obj_id_t oid, struct ioreq *req,
+			      char *data_buf, int num_recs,
+			      const char dkey[], const char akey[])
+{
+	int idx;
+	int i;
+
+	idx = 0;
+	print_message("Insert %d 4k records at index:0\n", num_recs);
+	for (i = 0; i < num_recs; i++) {
+		insert_single_with_rxnr(dkey, akey, idx, data_buf, 4096,
+					1, DAOS_TX_NONE, req);
+		idx++;
+		/**
+		 * Prevent records from coalescing on aggregation,
+		 * necessary to get a large amount of BIO reads
+		 * during aggregation of nvme extents
+		 */
+		idx += 1;
+	}
+	idx = 1;
+	print_message("Insert %d 4k records at index:1 (same key)\n",
+		      num_recs);
+	for (i = 0; i < num_recs; i++) {
+		insert_single_with_rxnr(dkey, akey, idx, data_buf, 4096,
+					1, DAOS_TX_NONE, req);
+		idx++;
+		/** Prevent records from coalescing on aggregation */
+		idx += 1;
+	}
+}
+
+/**
+ * Test max I/O threshold with aggregation (originally 512, bumped to 4k
+ * in DAOS-4269)
+ */
+static void
+max_io_agg_threshold(void **state)
+{
+	test_arg_t	*arg = *state;
+	char		*data_buf;
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+
+	oid = dts_oid_gen(dts_obj_class, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+
+	D_ALLOC(data_buf, IO_SIZE_NVME);
+	assert_non_null(data_buf);
+	dts_buf_render(data_buf, IO_SIZE_NVME);
+
+	/**
+	 * Insert 260 records twice under the same akey. This should result in
+	 * ~520 nvme segments being read for aggregation, which is larger
+	 * than the default max I/O threshold of 512. The max iops have now
+	 * been bumped to 4k, and anything over the threshold should not fail
+	 * and should be queued. Then insert 4100 records to test new threshold.
+	 */
+	max_io_agg_threshold_internal(oid, &req, data_buf, 260,
+				      "dkey_iomax", "akey_iomax");
+	max_io_agg_threshold_internal(oid, &req, data_buf, 2050,
+				      "dkey_iomax_4k", "akey_iomax_4k");
+
+	/** Wait >90 sec for aggregation to be triggered */
+	print_message("100 sec wait for aggregation to be triggered\n");
+	sleep(100);
+
+	D_FREE(data_buf);
+	ioreq_fini(&req);
+}
```

Comment:
I think it's a good manual test case, but it can't be part of CI test cases, because it doesn't detect failure at all. (No matter if the failure triggered by aggregation or not, this test will pass anyway, right?).

To make a auto test, I think we can add a VOS unit test which call bio_readv() or bio_writev() with more than 4k bio_iovs directly.

Reply:
I'm working on adding a test verification to check the blobstore health data to verify there were no bio read errors, I think that would be enough to validate this test. thoughts? I think adding a dc_mgmt call for all of the health query commands would be good anyways to validate all of the daos tests and future nvme recovery tests that need to be added for the faulty device detection feature.

Not sure if I will be able to get to the VOS unit test, but maybe we can put it on the backlog.

Reply:
Sure, that verify method looks good to me. I agree that we can put the VOS unit test in backlog.